crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::panic;
+use std::{sync::Arc, time::Duration};
+
+use broadcast::Receiver;
+use futures::stream::FuturesOrdered;
+use sui_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest},
+    committee::StakeUnit,
+    error::{SuiError, SuiResult},
+    messages::TransactionEffects,
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+#[cfg(test)]
+pub(crate) mod tests;
+
+const TASKS_PER_CORE: usize = 1;
+
+pub struct CheckpointExecutor {
+    mailbox: Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_executed_checkpoint: Option<CheckpointSequenceNumber>,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        mailbox: Receiver<VerifiedCheckpoint>,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        let highest_executed_checkpoint =
+            checkpoint_store.get_highest_executed_checkpoint_seq_number()?;
+
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            highest_executed_checkpoint,
+            authority_state,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        let mut finished: FuturesOrdered<JoinHandle<(u64, VerifiedCheckpoint)>> =
+            FuturesOrdered::new();
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
```

Comment:
I'd probably recommend not having this be a limit based on the number of cores and instead have this be a limit that is configurable via config (perhaps with a default based on the number of cores).

Reply:
Sounds good. Will followup later with a pr to add a config